src/blockchain/bitcoin_rpc.rs

```diff
@@ -0,0 +1,392 @@
+    json::{
+        ImportMultiOptions, ImportMultiRequest, ImportMultiRequestScriptPubkey,
+        ImportMultiRescanSince,
+    },
+    Client, RpcApi,
+};
+
+use super::*;
+use crate::database::{BatchDatabase, BatchOperations, DatabaseUtils};
+use crate::error::Error;
+use crate::types::{ScriptType, TransactionDetails, UTXO};
+
+pub struct BitcoinRpcBlockchain(Option<Client>);
+
+impl BitcoinRpcBlockchain {
+    // FIXME: why can I just say BitcoinRpcBlockchain(Some(client))?
+    pub fn new(client: Option<Client>) -> Self {
+        Self(client)
+    }
+    async fn index_is_synced<D: BatchDatabase + DatabaseUtils>(
+        &mut self,
+        index: u32,
+        database: &mut D,
+    ) -> Result<bool, Error> {
+        let networks = vec![Network::Bitcoin, Network::Testnet, Network::Regtest];
+        let path = DerivationPath::from(vec![ChildNumber::Normal { index }]);
+        let last_script = database
+            .get_script_pubkey_from_path(ScriptType::External, &path)?
+            .unwrap();
+        let synced = networks
+            .iter()
+            .map(|network| {
+                let address = Address::from_script(&last_script, *network).unwrap();
+                let response = self.0.as_mut().unwrap().get_address_info(&address);
+                if let Ok(result) = response {
+                    if let Some(is_watchonly) = result.is_watchonly {
+                        return is_watchonly;
+                    }
+                }
+                false
+            })
+            .any(|x| x);
+        Ok(synced)
+    }
+
+    async fn needs_sync<D: BatchDatabase + DatabaseUtils>(
+        &mut self,
+        database: &mut D,
+    ) -> Result<(bool, bool), Error> {
+        let synced = match database.get_last_index(ScriptType::External)? {
+            None => false,
+            Some(last_index) => self.index_is_synced(last_index, database).await?,
+        };
+        let needs_rescan = self.index_is_synced(0, database).await?;
+        Ok((!synced, needs_rescan))
+    }
+
+    async fn importmulti<D: BatchDatabase + DatabaseUtils>(
+        &mut self,
+        rescan_since: Option<u64>,
+        database: &mut D,
+    ) -> Result<(), Error> {
+        let (timestamp, rescan) = match rescan_since {
+            Some(timestamp) => (ImportMultiRescanSince::Timestamp(timestamp), Some(true)),
+            None => (ImportMultiRescanSince::Now, None),
+        };
+        self.0.as_mut().unwrap().import_multi(
+            &database
+                .iter_script_pubkeys(Some(ScriptType::External))?
+                .iter()
+                .map(|script_pubkey| ImportMultiRequest {
+                    script_pubkey: Some(ImportMultiRequestScriptPubkey::Script(&script_pubkey)),
+                    watchonly: Some(true),
+                    timestamp,
+                    ..Default::default()
+                })
+                .collect::<Vec<_>>(),
+            Some(&ImportMultiOptions { rescan }),
+        )?;
```

Comment:
Here I export all our scripts. It would be better if we could, say, export our last 100 scripts. Then check if the previous 100 scripts need exporting, etc.

Reply:
This still isn't ideal, but I think we can iterate on this.